### PR TITLE
Create initial Python bindings for c10d

### PR DESCRIPTION
**Pull request by pietern**
Pending:

* [x] Only compile c10d when compiling with CUDA support
* [x] Add compile time define for c10d to mask out c10d hooks in `torch/csrc/Module.cpp`

### CODE REVIEW

**Comment by pietern** [comment]
This includes #8206, so will rebase after that is merged.

**Comment by pietern** [comment]
There is some weirdness going on with the generated Caffe2 CMake files when cuDNN is found.

```
00:06:47.364 CMake Error at CMakeLists.txt:55 (add_library):
00:06:47.364   Target "c10d" links to target "caffe2::cudnn" but the target was not found.
00:06:47.364   Perhaps a find_package() call is missing for an IMPORTED target, or an
00:06:47.364   ALIAS target is missing?
```

**Review by teng-li — COMMENTED** [review]
Looks great overall

File: test/test_c10d.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+
+class StoreTestBase(object):
+    def _create_store(self, i):
+        raise RuntimeError("implement this")
+
+    def _test_set_get(self, fs):
+        fs.set("key0", "value0")
+        fs.set("key1", "value1")
+        fs.set("key2", "value2")
+        self.assertEqual(b"value0", fs.get("key0"))
```
[the remark this answers is not in the document]
Comment (apaszke) [reply test/test_c10d.py]: I think bytes are better, since it might contain binary data that isn't a valid string encoding, and that would throw.

File: test/test_c10d.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+        self.assertEqual(b"value1", fs.get("key1"))
+        self.assertEqual(b"value2", fs.get("key2"))
+
+    def test_set_get(self):
+        self._test_set_get(self._create_store())
+
+
+class FileStoreTest(TestCase, StoreTestBase):
+    def setUp(self):
+        self.file = tempfile.NamedTemporaryFile()
```
Comment (apaszke) [inline test/test_c10d.py]: You need to delete the file as well.

File: test/test_c10d.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    def _create_store(self):
+        return c10d.FileStore(self.file.name)
+
+
+class TCPStoreTest(TestCase, StoreTestBase):
+    def _create_store(self):
+        return c10d.TCPStore(TCP_ADDR, TCP_PORT, True)
+
+
+class ProcessGroupTestBase(object):
```
Comment (apaszke) [inline test/test_c10d.py]: This looks unused

File: test/test_c10d.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+    def setUpClass(cls):
+        for attr in dir(cls):
+            if attr.startswith('test'):
+                fn = getattr(cls, attr)
+                setattr(cls, attr, cls.join_or_run(fn))
+
+    def setUp(self):
+        self.rank = self.MAIN_PROCESS_RANK
+        self.size = 16
+        self.file = tempfile.NamedTemporaryFile()
```
Comment (apaszke) [inline test/test_c10d.py]: This needs to be deleted

File: test/test_c10d.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+        self.store = c10d.FileStore(self.file.name)
+        self.processes = []
+        for rank in range(int(self.size)):
+            self.processes.append(self._spawn_process(rank))
```
Comment (apaszke) [inline test/test_c10d.py]: nit: list comprehension?

File: test/test_c10d.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+        getattr(self, self.id().split(".")[2])()
+        sys.exit(0)
+
+    def _join_processes(self, fn):
+        timeout = get_timeout(self.id())
+        for p in self.processes:
+            p.join(timeout)
+
+    def test_broadcast_ops(self):
+        pg = c10d.ProcessGroupGloo(self.store, self.rank, self.size)
```
Comment (apaszke) [inline test/test_c10d.py]: I feel a bit uneasy having `store` be inherited via `fork`. It might lead to weird conditions in the future.

File: test/test_c10d.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+
+        def broadcast(xs, rootRank, rootTensor):
+            opts = c10d.BroadcastOptions()
+            opts.rootRank = rootRank
+            opts.rootTensor = rootTensor
+            work = pg.broadcast(xs, opts)
```
Comment (apaszke) [inline test/test_c10d.py]: This isn't a very friendly Python API. Can we make it consistent with how the distributed API looks like? `pg.broadcast(xs, root=X)`

File: test/test_c10d.py [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+    @classmethod
+    def setUpClass(cls):
+        for attr in dir(cls):
+            if attr.startswith('test'):
+                fn = getattr(cls, attr)
+                setattr(cls, attr, cls.join_or_run(fn))
+
+    def setUp(self):
+        self.rank = self.MAIN_PROCESS_RANK
+        self.size = 16
```
Comment (apaszke) [inline test/test_c10d.py]: There might be issues with too large sizes if we are going to test the NCCL backend (which needs at most a single rank per GPU).

**Review by apaszke — COMMENTED** [review]
Some comments to the tests

File: setup.py [hunk 9]
```diff
@@ -633,6 +635,8 @@ def run(self):
     CAFFE2_LIBS.extend(['-Wl,--no-as-needed', os.path.join(lib_path, 'libcaffe2_gpu.so'), '-Wl,--as-needed'])
 THD_LIB = os.path.join(lib_path, 'libTHD.a')
 NCCL_LIB = os.path.join(lib_path, 'libnccl.so.1')
+C10D_LIB = os.path.join(lib_path, 'libc10d.a')
+C10D_GLOO_LIB = os.path.join(lib_path, 'libc10d_gloo.a')
```
[the remark this answers is not in the document]
Comment (pietern) [reply setup.py]: Deferring to a later commit, don't think it is imperative to do this.

File: setup.py [hunk 10]
```diff
@@ -115,7 +115,7 @@
 from tools.setup_helpers.generate_code import generate_code
 from tools.setup_helpers.ninja_builder import NinjaBuilder, ninja_build_ext
 from tools.setup_helpers.dist_check import WITH_DISTRIBUTED, \
-    WITH_DISTRIBUTED_MW, WITH_GLOO_IBVERBS
+    WITH_DISTRIBUTED_MW, WITH_GLOO_IBVERBS, WITH_C10D
```
[the remark this answers is not in the document]
Comment (pietern) [reply setup.py]: WITH_DISTRIBUTED is still OK on Windows and without CUDA. Using a separate flag avoids conflating the existing meaning with the subset of environments that c10d will build in.

File: torch/csrc/c10d/init.cpp [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+#include <c10d/FileStore.hpp>
+#include <c10d/ProcessGroup.hpp>
+#include <c10d/ProcessGroupGloo.hpp>
+#include <c10d/TCPStore.hpp>
+
+#include "torch/csrc/Exceptions.h"
+#include "torch/csrc/utils/object_ptr.h"
+#include "torch/csrc/utils/pybind.h"
+
+namespace py = pybind11;
```
[the remark this answers is not in the document]
Comment (pietern) [reply torch/csrc/c10d/init.cpp]: Thanks.

File: torch/csrc/c10d/init.cpp [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+
+  auto store =
+      shared_ptr_class_<::c10d::Store>(module, "Store")
+          // Convert from std::string to std::vector<uint8>.
+          .def(
+              "set",
+              [](::c10d::Store& store,
+                 const std::string& key,
+                 const std::string& value) {
+                std::vector<unsigned char> value_(value.begin(), value.end());
```
[the remark this answers is not in the document]
Comment (pietern) [reply torch/csrc/c10d/init.cpp]: Good call. I thought I tried and failed but it works after all.

File: test/test_c10d.py [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+
+
+if not c10d.is_available():
+    print('c10d not available, skipping tests')
+    sys.exit(0)
+
+
+class StoreTestBase(object):
+    def _create_store(self, i):
+        raise RuntimeError("implement this")
```
[the remark this answers is not in the document]
Comment (pietern) [reply test/test_c10d.py]: Done.

File: test/test_c10d.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (pietern) [reply test/test_c10d.py]: Yep, we don't know it's valid UTF-8

File: test/run_test.py [hunk 14]
```diff
@@ -17,6 +17,7 @@
 TESTS = [
```
[the remark this answers is not in the document]
Comment (pietern) [reply test/run_test.py]: It has a shebang so my editor did that by itself (and I think stuff with shebang should have 755)

[the remark this answers is not in the document]
Comment (pietern) [reply test/run_test.py]: I'll revert it though

File: test/test_c10d.py [hunk 15]
```diff
@@ -0,0 +1,186 @@
+import math
+import multiprocessing
+import sys
+import tempfile
+import unittest
+from functools import wraps
+
+import torch
+import torch.c10d as c10d
```
[the remark this answers is not in the document]
Comment (pietern) [reply test/test_c10d.py]: c10 will also be top level, so I prefer this to also be top level.

Then when we are at parity we can make torch.distributed use stuff from torch.c10d, and eventually perhaps move it again. I'm not married to any particular location but like to keep it separate from torch.distributed while it's a work in progress.

File: test/test_c10d.py (hunk 2 above)
Comment (pietern) [reply test/test_c10d.py]: No need, from the docs: "If delete is true (the default), the file is deleted as soon as it is closed."

File: test/test_c10d.py (hunk 3 above)
Comment (pietern) [reply test/test_c10d.py]: Removed

File: test/test_c10d.py (hunk 5 above)
Comment (pietern) [reply test/test_c10d.py]: Done

File: test/test_c10d.py (hunk 6 above)
Comment (pietern) [reply test/test_c10d.py]: Good call, changed

File: test/test_c10d.py (hunk 7 above)
Comment (pietern) [reply test/test_c10d.py]: Good point. I prefer we do this in a Python-side wrapper that prepares whatever opts and does the kwargs handling. And we can do that in a later change once we have some factory function to create a process group, where we can easily wrap it in something that does the prettifying of the API.

Filed #8254 for this.

File: test/test_c10d.py (hunk 8 above)
Comment (pietern) [reply test/test_c10d.py]: Changed to 4.

Only the gloo backend is tested in this PR. When more are added we can do some better factoring for code reuse like you have done in test/test_distributed.py.

**Comment by pietern** [comment]
Rebasing to include #8248 that hopefully fixes the `caffe2::cudnn` CMake issue.

File: torch/lib/c10d/CMakeLists.txt [hunk 16]
```diff
@@ -71,19 +86,38 @@ set(C10D_GLOO_SRCS
...
+  set_property(TARGET c10d_mpi PROPERTY CXX_STANDARD 11)
   target_include_directories(c10d_mpi PUBLIC ${MPI_INCLUDE_PATH})
   target_link_libraries(c10d_mpi PUBLIC c10d ${MPI_LIBRARIES})
+  copy_header(ProcessGroupMPI.hpp)
+  target_include_directories(c10d_mpi PRIVATE ${CMAKE_BINARY_DIR}/include)
+  install(TARGETS c10d_mpi ARCHIVE DESTINATION lib)
 endif()
 
-add_subdirectory(example)
+option(BUILD_EXAMPLES "Build examples" OFF)
```
Comment (teng-li) [inline torch/lib/c10d/CMakeLists.txt]: shall we turn them ON by default?

File: torch/lib/c10d/CMakeLists.txt [hunk 17]
```diff
@@ -71,19 +86,38 @@ set(C10D_GLOO_SRCS
...
+if(BUILD_EXAMPLES)
+  add_subdirectory(example)
+endif()
+
+option(BUILD_TEST "Build tests" OFF)
```
Comment (teng-li) [inline torch/lib/c10d/CMakeLists.txt]: Same here

**Review by teng-li — APPROVED** [review]

File: test/test_c10d.py (hunk 2 above)
Comment (apaszke) [reply test/test_c10d.py]: Well you never call `.close()` and I'm pretty sure it's not closed automatically in `__del__`. Those objects are usualy used as context managers, and are deleted once you exit the scope.

File: test/test_c10d.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (apaszke) [reply test/test_c10d.py]: Well there's never going to be `torch.c10`, so I'm not sure which top level do you have in mind. I  think that if we want to let people use process groups as first class objects in their code, then they should just go under `torch.distributed.<some_readable_name_here>`. `c10d` is certainly not a very nice name for a Python module

**Review by apaszke — APPROVED** [review]
